ext/TensorKitChainRulesCoreExt/constructors.jl

```diff
@@ -17,6 +17,23 @@ function ChainRulesCore.rrule(::typeof(Base.copy), t::AbstractTensorMap)
     return copy(t), copy_pullback
 end
 
+function ChainRulesCore.rrule(::typeof(TensorKit.copy_oftype), t::AbstractTensorMap,
+                              T::Type{<:Number})
+    project = ProjectTo(t)
+    copy_oftype_pullback(Δt) = NoTangent(), project(unthunk(Δt)), NoTangent()
+    return TensorKit.copy_oftype(t, T), copy_oftype_pullback
+end
+
+function ChainRulesCore.rrule(::typeof(TensorKit.permutedcopy_oftype), t::AbstractTensorMap,
+                              T::Type{<:Number}, p::Index2Tuple)
+    project = ProjectTo(t)
+    function permutedcopy_oftype_pullback(Δt)
+        invp = TensorKit._canonicalize(TupleTools.invperm(linearize(p)), t)
+        return project(TensorKit.permutedcopy_oftype(unthunk(Δt), scalartype(t), invp))
```

Comment:
This doesn't work if `T` (which should coincide with `scalartype(Δt)`) is complex and `scalartype(t)` is real. This shouldn't really happen in the case where this function is being used (namely in the factorisations, where the purpose is to go from `Integer` or `Rational` types to floating point types, but not to change between real and complex). This triggers the question: should we just modify the tests, or make this work in general? There is no direct functionality to combine taking the real part and permuting.

Reply:
Oh I hadn't see your comment.